File: 1558-course-schedule-iv/1558-course-schedule-iv.py
class Solution:
    def dfs (self, node, adj, visited, prerequisites, start):
        visited [node] = True

        if start != -1:
            prerequisites.add ((start, node))

        for neighbor in adj [node]:
            if not visited [neighbor]:
                self.dfs (neighbor, adj, visited, prerequisites, start)

    def checkIfPrerequisite (self, numCourses : int, prerequisites : List [List [int]], queries : List [List [int]]) -> List [bool]:

        adj = [[] for _ in range (numCourses)]
        for u, v in prerequisites:
            adj [u].append (v)

        prerequisites = set ()
        for i in range (numCourses):
            visited = [False] * numCourses
            self.dfs (i, adj, visited, prerequisites, i)

        return [tuple (query) in prerequisites for query in queries]

File: 1558-course-schedule-iv/1558-course-schedule-iv.java
class Solution {
    public List <Boolean> checkIfPrerequisite (int numCourses, int [] [] prerequisites, int [] [] queries) {
        List <Integer> [] adj = new ArrayList [numCourses];
        boolean [] [] prereq = new boolean [numCourses] [numCourses];
        int [] inDegree = new int [numCourses];
        
        for (int i = 0; i < numCourses; i++) {
            adj [i] = new ArrayList <> ();
        }
        
        for (int[] edge : prerequisites) {
            int a = edge [0], b = edge [1];
            adj [a].add (b);
            prereq [b] [a] = true;
            inDegree [b]++;
        }
        
        Queue <Integer> q = new LinkedList <> ();
        for (int i = 0; i < numCourses; i++) {
            if (inDegree [i] == 0) q.add (i);
        }
        
        while (!q.isEmpty ()) {
            int u = q.poll ();
            for (int v : adj [u]) {
                for (int i = 0; i < numCourses; i++) {
                    if (prereq [u] [i]) prereq [v] [i] = true;
                }
                if (--inDegree [v] == 0) q.add (v);
            }
        }
        
        List <Boolean> ans = new ArrayList <> ();
        for (int [] query : queries) {
            int u = query [0], v = query [1];
            ans.add (prereq [v] [u]);
        }
        return ans;
    }
}